receiver/otlpreceiver/config.go

```diff
@@ -0,0 +1,144 @@
+
+	// Keepalive anchor for all the settings related to keepalive.
+	Keepalive *serverParametersAndEnforcementPolicy `mapstructure:"keepalive,omitempty"`
+
+	// MaxRecvMsgSizeMiB sets the maximum size (in MiB) of messages accepted by the server.
+	MaxRecvMsgSizeMiB uint64 `mapstructure:"max_recv_msg_size_mib,omitempty"`
+
+	// MaxConcurrentStreams sets the limit on the number of concurrent streams to each ServerTransport.
+	MaxConcurrentStreams uint32 `mapstructure:"max_concurrent_streams,omitempty"`
+}
+
+type serverParametersAndEnforcementPolicy struct {
+	ServerParameters  *keepaliveServerParameters  `mapstructure:"server_parameters,omitempty"`
+	EnforcementPolicy *keepaliveEnforcementPolicy `mapstructure:"enforcement_policy,omitempty"`
+}
+
+// keepaliveServerParameters allow configuration of the keepalive.ServerParameters.
+// The same default values as keepalive.ServerParameters are applicable and get applied by the server.
+// See https://godoc.org/google.golang.org/grpc/keepalive#ServerParameters for details.
+type keepaliveServerParameters struct {
+	MaxConnectionIdle     time.Duration `mapstructure:"max_connection_idle,omitempty"`
+	MaxConnectionAge      time.Duration `mapstructure:"max_connection_age,omitempty"`
+	MaxConnectionAgeGrace time.Duration `mapstructure:"max_connection_age_grace,omitempty"`
+	Time                  time.Duration `mapstructure:"time,omitempty"`
+	Timeout               time.Duration `mapstructure:"timeout,omitempty"`
+}
+
+// keepaliveEnforcementPolicy allow configuration of the keepalive.EnforcementPolicy.
+// The same default values as keepalive.EnforcementPolicy are applicable and get applied by the server.
+// See https://godoc.org/google.golang.org/grpc/keepalive#EnforcementPolicy for details.
+type keepaliveEnforcementPolicy struct {
+	MinTime             time.Duration `mapstructure:"min_time,omitempty"`
+	PermitWithoutStream bool          `mapstructure:"permit_without_stream,omitempty"`
+}
+
+func (rOpts *Config) buildOptions() (opts []Option, err error) {
+	tlsCredsOption, hasTLSCreds, err := ToOTLPReceiverServerOption(rOpts.TLSCredentials)
+	if err != nil {
+		return opts, fmt.Errorf("error initializing OTLP receiver %q TLS Credentials: %v", rOpts.NameVal, err)
+	}
+	if hasTLSCreds {
+		opts = append(opts, tlsCredsOption)
+	}
+	if len(rOpts.CorsOrigins) > 0 {
+		opts = append(opts, WithCorsOrigins(rOpts.CorsOrigins))
+	}
+
+	grpcServerOptions := rOpts.grpcServerOptions()
+	if len(grpcServerOptions) > 0 {
+		opts = append(opts, WithGRPCServerOptions(grpcServerOptions...))
+	}
+
+	return opts, err
+}
+
+func (rOpts *Config) grpcServerOptions() []grpc.ServerOption {
+	var grpcServerOptions []grpc.ServerOption
+	if rOpts.MaxRecvMsgSizeMiB > 0 {
+		grpcServerOptions = append(grpcServerOptions, grpc.MaxRecvMsgSize(int(rOpts.MaxRecvMsgSizeMiB*1024*1024)))
+	}
+	if rOpts.MaxConcurrentStreams > 0 {
+		grpcServerOptions = append(grpcServerOptions, grpc.MaxConcurrentStreams(rOpts.MaxConcurrentStreams))
```

Comment:
This parameter gets applied as an HTTP/2 setting, see https://tools.ietf.org/html/rfc7540#section-6.5.2
```
   SETTINGS_MAX_CONCURRENT_STREAMS (0x3):  Indicates the maximum number
      of concurrent streams that the sender will allow.  This limit is
      directional: it applies to the number of streams that the sender
      permits the receiver to create.  Initially, there is no limit to
      this value.  It is recommended that this value be no smaller than
      100, so as to not unnecessarily limit parallelism.
```
and the "stream" here refers to the HTTP/2 stream https://developers.google.com/web/fundamentals/performance/http2#streams_messages_and_frames

So seems applicable given that we allow multiple concurrent requests.

Reply:
Yes, I would think so too, but it would be good to verify this actually works as we imagine it should (i.e. limits the number of concurrent unary requests). Can you add a TODO near the config option that says "need to be tested".

Reply:
Added a TODO comment.